### PR TITLE
Fix #4395: Handle ## when it is an Ident

### LINKED ISSUE #4395: Compiler crash: NoDenotation.owner
```scala
object Test {
  val res0 = ##
}
```
```
java.lang.AssertionError: NoDenotation.owner
	at dotty.tools.dotc.core.SymDenotations$NoDenotation$.owner(SymDenotations.scala:1898)
	at dotty.tools.backend.jvm.DottyBackendInterface$$anon$9.owner(DottyBackendInterface.scala:749)
	at dotty.tools.backend.jvm.DottyBackendInterface$$anon$9.owner(DottyBackendInterface.scala:653)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genCallMethod(BCodeBodyBuilder.scala:1113)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genScalaHash(BCodeBodyBuilder.scala:1171)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genPrimitiveOp(BCodeBodyBuilder.scala:237)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:711)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:337)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genStat(BCodeBodyBuilder.scala:77)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.$anonfun$genBlock$1(BCodeBodyBuilder.scala:860)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.$anonfun$genBlock$1$adapted(BCodeBodyBuilder.scala:860)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genNormalBlock$1(BCodeBodyBuilder.scala:860)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genBlock(BCodeBodyBuilder.scala:892)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:408)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.emitNormalMethodBody$1(BCodeSkelBuilder.scala:602)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genDefDef(BCodeSkelBuilder.scala:635)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:506)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.$anonfun$gen$1(BCodeSkelBuilder.scala:508)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.$anonfun$gen$1$adapted(BCodeSkelBuilder.scala:508)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:508)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:116)
	at dotty.tools.backend.jvm.GenBCodePipeline$Worker1.visit(GenBCode.scala:224)
	at dotty.tools.backend.jvm.GenBCodePipeline$Worker1.run(GenBCode.scala:191)
	at dotty.tools.backend.jvm.GenBCodePipeline.buildAndSendToDisk(GenBCode.scala:421)
	at dotty.tools.backend.jvm.GenBCodePipeline.run(GenBCode.scala:387)
	at dotty.tools.backend.jvm.GenBCode.run(GenBCode.scala:69)
	at dotty.tools.dotc.core.Phases$Phase.$anonfun$runOn$1(Phases.scala:295)
```

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/InterceptedMethods.scala [hunk 1]
```diff
@@ -54,6 +38,16 @@ class InterceptedMethods extends MiniPhase {
     else tree
   }
 
+  override def transformIdent(tree: tpd.Ident)(implicit ctx: Context): Tree = {
+    if (tree.symbol.isTerm && (defn.Any_## eq tree.symbol.asTerm)) {
+      val cls = ctx.owner.ownersIterator.find(_.isClass).get.asClass
+      val rewrite = poundPoundValue(This(cls))
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/InterceptedMethods.scala]: I think what you want is `tpd.desugarIdentPrefix`

File: compiler/src/dotty/tools/dotc/transform/InterceptedMethods.scala [hunk 2]
```diff
@@ -54,6 +38,16 @@ class InterceptedMethods extends MiniPhase {
     else tree
   }
 
+  override def transformIdent(tree: tpd.Ident)(implicit ctx: Context): Tree = {
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/InterceptedMethods.scala]: Can you factor out the common code between `transformIdent` and `transformSelect`

Comment (allanrenucci) [reply compiler/src/dotty/tools/dotc/transform/InterceptedMethods.scala]: What about something like:
```scala
private def transformRefTree(tree: RefTree)(implicit ctx: Context): Tree =
  if (tree.symbol.isTerm && (defn.Any_## eq tree.symbol.asTerm)) {
    val qual = tree match
      case id: Ident =>tpd.desugarIdentPrefix(id)
      case sel: Select => sel.qual
    }
    val rewrite = poundPoundValue(qual)
    ctx.log(s"$phaseName rewrote $tree to $rewrite")
    rewrite
  }
  else tree

override def transformIdent(tree: tpd.Ident)(implicit ctx: Context) =
  transformPoundPound(tree)

override def transformSelect(tree: tpd.Select)(implicit ctx: Context): Tree =
  transformPoundPound(tree)
```

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/InterceptedMethods.scala]: Done

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/InterceptedMethods.scala]: Though it added a line of code

**Pull request merged**